### ISSUE TITLE
Expose "process exited" hook on ProcessProtocol, turning current processEnded into user-overriddable behavior

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10883]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#1291|
|Type|enhancement|
|Created|2005-10-28 01:06:44Z|
|Branch|https://github.com/twisted/twisted/tree/process-exited-hook-1291-3|

<details><summary>Searchable metadata</summary>

```
trac-id__1291 1291
type__enhancement enhancement
reporter__itamarst itamarst
priority__normal normal
milestone__ 
branch__branches_process_exited_hook_1291_3 branches/process-exited-hook-1291-3
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1130461604000000 1130461604000000
changetime__1217104678000000 1217104678000000
version__ 
owner__ 
cc__exarkun cc__itamarst cc__jknight cc__radix
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
&lt;itamar> right now we have:
&lt;itamar> 1. fd closed
&lt;itamar> 2. fds all closed + process exited
&lt;itamar> I propose
&lt;itamar> 3. process exited

exarkun improved this with:
&lt;exarkun> ok how about this
&lt;exarkun> add the process exited hook
&lt;exarkun> the default implementation should call processEnded when appropriate
&lt;exarkun> and processEnded shouldn't be part of the official api

&lt;exarkun> there should be a real interface for ProcessProtocol
&lt;exarkun> and test coverage for edge cases like this
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

[#2774](https://github.com/twisted/twisted/issues/2774) was closed as a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [24015]) Branching to 'process-exited-hook-1291'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [24070]) Branching to 'process-exited-hook-1291-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Okay.

Build results:

http://buildbot.twistedmatrix.com/boxes-supported?num_builds=1&branch=/branches/process-exited-hook-1291-2

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_7"></a>|@radix commented|
|-|-|

Do you know what's up with this?

http://buildbot.twistedmatrix.com/builders/server-2k8-x86-py2.5-select/builds/189/steps/select/logs/problems

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_9"></a>|@radix set owner to @exarkun|
|-|-|

1. You're checking if the return value of _getReason is None but it doesn't look like you need to.

1. the handling of processExited is different between the win32 and posix cases: if you set it to None on win32, it'll be ignored. if you set it to None on posix, it'll blow up. Just pointing this out because I thought it was slightly strange that you used two different ways of optionally calling processExited.

1. So... wait a second. We used to call processEnded immediately on process exit on Windows? Oi.

1. This windows shit is horrible. Will the change you made to call addPollableResource earlier result in any significant performance differences? it looks like maybe not, since the polling is all shared in one callLater and all the FDs are being polled anyway. I guess it just increases N.

1. processEnded needs to be documented (in t.i.i) as being called only when the FDs are closed and the process has exited.

1. processExited needs to be added to t.i.i.

1. doc/core/howto/processes.xhtml should be updated to talk about processExited instead of processEnded.

1. Are we going to emit a warning when processExited isn't available? Do we need to change ProcessProtocol in any way? Should processEnded be deprecated?

Ok, that should be enough for now. I haven't even looked at the tests yet.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [24122]) remove superfluous None check

refs [#10883](https://github.com/twisted/twisted/issues/10883)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [24126]) add processExited to IProcessProtocol

refs [#10883](https://github.com/twisted/twisted/issues/10883)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [24175]) Branching to 'process-exited-hook-1291-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun removed owner|
|-|-|

Replying to [radix](#note_9):
> 1. You're checking if the return value of _getReason is None but it doesn't look like you need to.

Removed.

> 1. the handling of processExited is different between the win32 and posix cases: if you set it to None on win32, it'll be ignored. if you set it to None on posix, it'll blow up. Just pointing this out because I thought it was slightly strange that you used two different ways of optionally calling processExited.

Factored `processExited` calling into a shared base class, along with a couple other pieces of code.

> 1. So... wait a second. We used to call processEnded immediately on process exit on Windows? Oi.

I don't think so.  It counted three handle closes and only then started checking to see if the process had exited.

> 1. This windows shit is horrible. Will the change you made to call addPollableResource earlier result in any significant performance differences? it looks like maybe not, since the polling is all shared in one callLater and all the FDs are being polled anyway. I guess it just increases N.

Yea, it's just one more check (at least) ten times per second.  Four per process, now, instead of three.  It probably has some performance difference, but I wouldn't expect a lot.  I don't know how much, we don't have any benchmarks for process support on Windows.

> 1. processEnded needs to be documented (in t.i.i) as being called only when the FDs are closed and the process has exited.

Done

> 1. processExited needs to be added to t.i.i.

Done

> 1. doc/core/howto/processes.xhtml should be updated to talk about processExited instead of processEnded.

Done, although I find `process.xhtml` to be pretty bad in general still (since it has wrong information and for some reason decides to document most things twice in a subtly different way).

> 1. Are we going to emit a warning when processExited isn't available? Do we need to change ProcessProtocol in any way? Should processEnded be deprecated?

I added a `DeprecationWarning` when `processExited` isn't implemented, and I added `processExited` to `ProcessProtocol`.  I didn't do anything to `processEnded`, I don't think we can deprecate it until we untangle the PTY message.

> Ok, that should be enough for now. I haven't even looked at the tests yet.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_14"></a>|@radix set owner to @exarkun|
|-|-|



1.

```
- <code>outReceived</code> ought to simply accumulate the data and put off
+  <code>outReceived</code> must ought to simply accumulate the data and put off
```

This doesn't look like a good change. I guess you meant to delete the "ought to"?

2. I don't think you need to document processExited the second time in process.xhtml. It seems that section is only for the things which are different from the "I don't care about file descriptors" mode which is the first list of things. Yeah, this document sucks horribly.


Ok, I think I'm happy with this branch now. Feel free to fix these minor documentation nits, and assuming the builders are still passing, go ahead and merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [24390]) undo the start of an edit that I decided was too big to do in this branch

the recommendation to buffer everything until the process exits is naive at
best

refs [#10883](https://github.com/twisted/twisted/issues/10883)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [24391]) remove processExited from the 2nd copy of the `ProcessProtocol` documentation

refs [#10883](https://github.com/twisted/twisted/issues/10883)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set status to `closed`|
|-|-|

(In [24392]) Merge process-exited-hook-1291-3

Author: exarkun
Reviewer: radix
Fixes: [#10883](https://github.com/twisted/twisted/issues/10883)

Add `processExited` to `IProcessProtocol`, a callback which is invoked when
a child process exits.  This is distinct from `IProcessProtocol.processEnded`
which is only called after the child process exits and all of the file
descriptors associated with it have been closed.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
